### PR TITLE
[data_files] Only match separated split names

### PR DESCRIPTION
**Pull request by lhoestq**
As reported in https://github.com/huggingface/datasets/issues/4477, the current pattern matching to infer which file goes into which split is too permissive. For example a file "contest.py" would be considered part of a test split (it contains "test") and "seqeval.py" as well (it contains "eval").

In this PR I made the pattern matching more robust by only matching split names **between separators**. The supported separators are dots, dashes, spaces and underscores.

I updated the docs accordingly.

One detail about the tests: I had to update one test because it was using `PurePath.match` as a reference for globbing, but it doesn't support the `[..]` glob pattern. Therefore I added a `mock_fs` context manager that can be used to easily define a dummy filesystem with certain files in it and run pattern matching tests. Its code comes mostly from test_streaming_download_manager.py

Close https://github.com/huggingface/datasets/issues/4477

### CODE REVIEW

**Comment by lhoestq** [comment]
I ran a script to find affected datasets. Adding "testing" and "evaluation" fixes all of of them except one:
- projecte-aina/cat_manynames:	human_annotated_testset.tsv

Let me open a PR on their repository to fix it

**Review by mariosasko — APPROVED** [review]
All looks good now. Thanks!

File: tests/test_data_files.py [hunk 1]
```diff
@@ -491,6 +493,69 @@ def test_DataFilesDict_from_hf_local_or_remote_hashing(text_file):
...
+@contextmanager
+def mock_fs(file_paths: List[str]):
+    """context manager to set up a mock:// filesystem in sfspec containing the provided files"""
+
+    dir_paths = {file_path.rsplit("/")[0] for file_path in file_paths if "/" in file_path}
+    fs_contents = [{"name": dir_path, "type": "directory"} for dir_path in dir_paths] + [
+        {"name": file_path, "type": "file", "size": 10} for file_path in file_paths
+    ]
+
+    class DummyTestFS(AbstractFileSystem):
```
Comment (albertvillanova) [inline tests/test_data_files.py:505]: Awesome that you use a mock file system. I personally think this goes in the right direction.

File: tests/test_data_files.py [hunk 2]
```diff
@@ -491,6 +493,69 @@ def test_DataFilesDict_from_hf_local_or_remote_hashing(text_file):
...
+                path,
+                mode,
+                block_size,
+                autocommit,
+                cache_options=cache_options,
+                **kwargs,
+            )
+
+    with patch.dict(fsspec.registry.target, {"mock": DummyTestFS}):
+        yield DummyTestFS()
```
Comment (albertvillanova) [inline tests/test_data_files.py:556]: Just a comment (feel free to ignore it): you use here `unittest.mock.patch`, but you could use `pytest.monkeypatch` instead.

File: tests/test_data_files.py [hunk 3]
```diff
@@ -491,6 +493,69 @@ def test_DataFilesDict_from_hf_local_or_remote_hashing(text_file):
         assert Hasher.hash(data_files1) != Hasher.hash(data_files2)
 
 
+@contextmanager
+def mock_fs(file_paths: List[str]):
+    """context manager to set up a mock:// filesystem in sfspec containing the provided files"""
```
Comment (albertvillanova) [inline tests/test_data_files.py:498]: Typo
```suggestion
    """context manager to set up a mock:// filesystem in fsspec containing the provided files"""
```

File: tests/test_data_files.py [hunk 4]
```diff
@@ -541,25 +606,30 @@ def test_DataFilesDict_from_hf_local_or_remote_hashing(text_file):
         {"validation": "dev/dataset.txt"},
         # With other extensions
         {"train": "train.parquet", "test": "test.parquet", "validation": "valid.parquet"},
+        # With "dev" or "eval" without separators
+        {"train": "developers_list.txt"},
+        {"train": "data/seqeval_results.txt"},
```
Comment (albertvillanova) [inline tests/test_data_files.py:611]: Maybe also adding a test for "test": "contest.txt"?

File: tests/test_data_files.py [hunk 5]
```diff
@@ -541,25 +606,30 @@ def test_DataFilesDict_from_hf_local_or_remote_hashing(text_file):
...
+    ) as fs:
+
+        def resolver(pattern):
+            return [PurePath(file_path) for file_path in fs.glob(pattern) if fs.isfile(file_path)]
+
+        patterns_per_split = _get_data_files_patterns(resolver)
+        assert sorted(patterns_per_split.keys()) == sorted(data_file_per_split.keys())
+        for split, patterns in patterns_per_split.items():
+            matched = [file_path.as_posix() for pattern in patterns for file_path in resolver(pattern)]
+            assert matched == data_file_per_split[split]
```
Comment (albertvillanova) [inline tests/test_data_files.py:632]: Maybe better moving the context manager to the specific position where it is necessary?
```suggestion

    def resolver(pattern):
        with mock_fs(
            [file_path for split_file_paths in data_file_per_split.values() for file_path in split_file_paths]
        ) as fs:
            return [PurePath(file_path) for file_path in fs.glob(pattern) if fs.isfile(file_path)]

    patterns_per_split = _get_data_files_patterns(resolver)
    assert sorted(patterns_per_split.keys()) == sorted(data_file_per_split.keys())
    for split, patterns in patterns_per_split.items():
        matched = [file_path.as_posix() for pattern in patterns for file_path in resolver(pattern)]
        assert matched == data_file_per_split[split]
```

**Review by albertvillanova — COMMENTED** [review]
Awesome!! This will definitely make the user experience much more pleasant and aligned with their expectations.

File: tests/test_data_files.py (hunk 2 above)
Comment (lhoestq) [reply tests/test_data_files.py:556]: This function is a context manager that doesn't have access to the `monkeypatch` fixture of pytest, so I used `unittest.mock.patch` instead.

**Review by albertvillanova — APPROVED** [review]
Thank you!! Nice job.

File: tests/test_data_files.py (hunk 2 above)
Comment (albertvillanova) [reply tests/test_data_files.py:562]: @lhoestq again not important: but indeed you are not using the patching (neither `unittest.mock.patch` nor `pytest.monkeypatch`. You are just using the returned instance `DummyTestFS()`.

So I guess you could just remove the patching and the test will pass anyway.

Comment (albertvillanova) [reply tests/test_data_files.py:562]:
```suggestion
    yield DummyTestFS()
```

**Pull request merged**